firmware/gibibot_firmware/Core/Src/main.c

```diff
@@ -0,0 +1,950 @@
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+TIM_HandleTypeDef htim1;
+TIM_HandleTypeDef htim2;
+TIM_HandleTypeDef htim3;
+TIM_HandleTypeDef htim4;
+TIM_HandleTypeDef htim6;
+TIM_HandleTypeDef htim8;
+TIM_HandleTypeDef htim15;
+TIM_HandleTypeDef htim16;
+TIM_HandleTypeDef htim17;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_TIM1_Init(void);
+static void MX_TIM2_Init(void);
+static void MX_TIM3_Init(void);
+static void MX_TIM4_Init(void);
+static void MX_TIM15_Init(void);
+static void MX_TIM16_Init(void);
+static void MX_TIM17_Init(void);
+static void MX_TIM8_Init(void);
+static void MX_TIM6_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+typedef struct Command {
+  int16_t speed1;
+  int16_t speed2;
+  int16_t speed3;
+  uint16_t throwerSpeed;
+  uint16_t throwerAngle;
+  uint16_t throwerGrab;
+  uint16_t delimiter;
+} Command;
+
+typedef struct Feedback {
+  //int16_t speed1;
+  int16_t ballDetected;
+  int16_t speed2;
+  int16_t speed3;
+  uint16_t delimiter;
+} Feedback;
+int16_t IRsensor;
+typedef struct MotorControl {
+	int16_t gainP;
+	int16_t gainI;
+	int16_t gainD;
+	int32_t integral;
+	int16_t positionChange;
+	int16_t position;
+} MotorControl;
+
+MotorControl motorControl1 = {.position = 0, .positionChange = 0, .gainP = 1000, .gainI = 50, .gainD = 10, .integral = 0};
+MotorControl motorControl2 = {.position = 0, .positionChange = 0, .gainP = 1000, .gainI = 50, .gainD = 10, .integral = 0};
+MotorControl motorControl3 = {.position = 0, .positionChange = 0, .gainP = 1000, .gainI = 50, .gainD = 10, .integral = 0};
+
+Command command = {
+		.speed1 = 0,
+		.speed2 = 0,
+		.speed3 = 0,
+		.throwerSpeed = 0,
+		.throwerAngle = 6900,
+		.throwerGrab = 4800,
+		.delimiter = 0};
+
+volatile uint8_t isCommandReceived = 0;
+
+void CDC_On_Receive(uint8_t* buffer, uint32_t* length) {
+  if (*length == sizeof(Command)) {
+    memcpy(&command, buffer, sizeof(Command));
+
+    if (command.delimiter == 0xAAAA) {
+      isCommandReceived = 1;
+    }
+  }
+}
+
+int32_t ClampValue(int32_t val, int32_t min, int32_t max) {
+	if (val < min) {
+		return min;
+	}
+	if (val > max) {
+		return max;
+	}
+	return val;
+}
+
+uint16_t ClampValue2(uint16_t val, uint16_t min, uint16_t max) {
+	if (val < min) {
+		return min;
+	}
+	if (val > max) {
+		return max;
+	}
+	return val;
+}

```

Comment:
`ClampValue` and `ClampValue2` are not very good names.

Something like `ClampValueI32` and `ClampValueU16` would be better.